### ISSUE TITLE
Unable to register dockerd.exe service on Windows 11

### ISSUE THREAD

**Issue by Edstub207** [docker/for-win#13600]
### Description

After installing Docker on Windows 11 I'm unable to register the daemon as a service 

```
C:\Program Files\Docker\Docker\resources>dockerd --register-service
exec: "dockerd": cannot run executable found relative to current directory
```

### Reproduce

- Install Docker for Windows v4.21.1 on Windows 22H2
- Navigate to C:\Program Files\Docker\Docker\resources
- Verify dockerd.exe exists
- Open CMD window in admin mode
- CD into  C:\Program Files\Docker\Docker\resources
- Run `dockerd --register-service` (And it's variations such as dockerd.exe)

### Expected behavior

Docker service is installed

### docker version

```bash
Client:
 Cloud integration: v1.0.35
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:53:15 2023
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Desktop 4.21.1 (114176)
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       659604f
  Built:            Thu May 25 21:52:17 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    24.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-buildx.exe
  compose: Docker Compose (Docker Inc.)
    Version:  v2.19.1
    Path:     C:\Program Files\Docker\cli-plugins\docker-compose.exe
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-dev.exe
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.20
    Path:     C:\Program Files\Docker\cli-plugins\docker-extension.exe
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.6
    Path:     C:\Program Files\Docker\cli-plugins\docker-init.exe
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-sbom.exe
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-scan.exe
  scout: Command line tool for Docker Scout (Docker Inc.)
    Version:  0.16.1
    Path:     C:\Program Files\Docker\cli-plugins\docker-scout.exe

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 5.15.90.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 62.76GiB
 Name: docker-desktop
 ID: a4145cf6-f266-4799-81d3-b831fa32a1cb
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

### Diagnostics ID

0E2122C6-4F67-4C0A-BAEC-3B1A7F84BFE9/20230713174151

**Comment by lorenrh** [comment]
Hello!

What is the use case for this?

**Comment by Edstub207** [comment]
> Hello!
> 
> What is the use case for this? 

@lorenrh hi, we use it to ensure Docker is fully up and running on startup of the machine without a user needing to log in, which is required for docker desktop to be started.

This is because we use Docker on our build machines with automated pipelines that require docker.

It works on Windows 10.

**Comment by thaJeztah** [comment]
This error comes from the Go runtime, and IIRC this was for a security fix for Windows; https://github.com/golang/go/blob/e713d6f939c90eb599c1469d08bb5edd7de8a281/src/os/exec/exec.go#L1303

Not sure why it would trigger in this case though.


Does it work if you don't navigate to the Directory, and instead execute the binary using its full path?

**Comment by Edstub207** [comment]
> This error comes from the Go runtime, and IIRC this was for a security fix for Windows; https://github.com/golang/go/blob/e713d6f939c90eb599c1469d08bb5edd7de8a281/src/os/exec/exec.go#L1303
> 
> Not sure why it would trigger in this case though.
> 
> 
> Does it work if you don't navigate to the Directory, and instead execute the binary using its full path?

@thaJeztah interesting, thanks for the info. I'm currently AFK until the middle of next week, I can imagine manually creating the service would work. Are there any specific arguments or setup suggestions required if doing this manually?

I think I tried outside the directory but I can try again

**Comment by Edstub207** [comment]
@thaJeztah I can confirm backing out a directory up and doing the following worked (Obviously as an Admin CMD prompt session)

`C:\Program Files\Docker\Docker>resources\dockerd.exe --register-service`

**Comment by thaJeztah** [comment]
Thanks for confirming! I opened a pull-request that should hopefully fix this (I'm not on a Windows machine, so need to ask someone to try 😂). Given that there's a workaround, I'm not sure if it would qualify for backporting to existing releases, but if it  gets merged, it will likely land in the v25.0.0 release of the Docker Engine;

- https://github.com/moby/moby/pull/46091

**Comment by thaJeztah** [comment]
^^ github closed this because the fix was merged, but (for transparency) it wasn't released yet, and didn't ship yet 😅
